# Taller 1

**Fecha de entrega:** Miércoles 6 de agosto

Se tiene un conjunto de datos de 1500 registros con las siguientes variables:

- edad_cliente (años)
- tiempo_como_cliente_meses
- ingresos_mensuales (dolares)
- visitas_mensuales_sitio_web
- contactos (call-center)
- nivel_educativo (1:primaria, 2:secundaria, 3:tecnico-tecnologico, 4:universitario, 5 posgrado)
- uso_tarjeta_credito (1: Si, 0: No)
- participa_en_promociones (1: Si, 0: No)
- calificacion_satisfaccion (1-5, donde 1 es muy bajo y 5 es excelente)
- descuento_promedio_usado (%)
- ventas_cliente (Ventas promedio al cliente en dolares)


Se espera que el cientifico de datos presente el mejor modelo de regresión posible para predecir las ventas del cliente, con el respectivo informe ejecutivo para el gerente.
Se deben validar todos los supuestos de la regresión y que esta no presente problemas.


1. Cargar librerias

In [3]:
import numpy as np
import pandas as pd
import sklearn.linear_model as LinearRegression
import matplotlib.pyplot as plt
import statsmodels.api as sm
import scipy.stats as stats

2. Cargar datos

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
df=pd.read_csv('/content/drive/MyDrive/Aprendizaje Supervisado/dataset_regresion_ventas.csv')
df.head()
df.columns

Index(['edad_cliente', 'tiempo_como_cliente_meses', 'ingresos_mensuales',
       'visitas_mensuales_sitio_web', 'contactos', 'nivel_educativo',
       'uso_tarjeta_credito', 'participa_en_promociones',
       'calificacion_satisfaccion', 'descuento_promedio_usado',
       'ventas_cliente'],
      dtype='object')

In [5]:
train=df.sample(frac=0.7,random_state=200)
test=df.drop(train.index)

3. Regresion Lineal Train

In [32]:
yt=train['ventas_cliente']
xt=train[['edad_cliente','tiempo_como_cliente_meses','visitas_mensuales_sitio_web','descuento_promedio_usado']]

xt=np.array(xt)
yt=np.array(yt)

modelotrain=LinearRegression.LinearRegression()
modelotrain.fit(xt, yt)

print('Intercerpto(a):',modelotrain.intercept_)
print('Pendientes(b):',modelotrain.coef_)

Intercerpto(a): 152.8567011884595
Pendientes(b): [  2.60266071   1.67174481   7.07495136 344.71459861]


4. Modelo de regresion Train

In [33]:
x_constt=sm.add_constant(xt)
modelo_smt=sm.OLS(yt,x_constt).fit()
print(modelo_smt.summary())

residuales = modelo_smt.resid
print("\nResiduales:")
print(residuales)

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.375
Model:                            OLS   Adj. R-squared:                  0.373
Method:                 Least Squares   F-statistic:                     157.1
Date:                Thu, 07 Aug 2025   Prob (F-statistic):          2.92e-105
Time:                        01:30:23   Log-Likelihood:                -6358.3
No. Observations:                1050   AIC:                         1.273e+04
Df Residuals:                    1045   BIC:                         1.275e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.8567     13.920     10.981      0.0

5. Regresion Lineal Test

In [35]:
y=test['ventas_cliente']
x=test[['edad_cliente','tiempo_como_cliente_meses','visitas_mensuales_sitio_web','descuento_promedio_usado']]

x=np.array(x)
y=np.array(y)

modelotest=LinearRegression.LinearRegression()
modelotest.fit(x, y)

print('Intercerp(a):',modelotest.intercept_)
print('Pendientes(b):',modelotest.coef_)

Intercerp(a): 159.28546546936514
Pendientes(b): [  2.36215701   1.83418048   6.4814798  298.87969767]


6. Modelo de regresion lineal Test



In [36]:
x_const=sm.add_constant(x)
modelo_sm=sm.OLS(y,x_const).fit()
print(modelo_sm.summary())

residuales = modelo_sm.resid
print("\nResiduales:")
print(residuales)

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.370
Model:                            OLS   Adj. R-squared:                  0.364
Method:                 Least Squares   F-statistic:                     65.30
Date:                Thu, 07 Aug 2025   Prob (F-statistic):           1.98e-43
Time:                        01:30:38   Log-Likelihood:                -2726.0
No. Observations:                 450   AIC:                             5462.
Df Residuals:                     445   BIC:                             5483.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        159.2855     21.976      7.248      0.0

**Informe Ejecutivo**

En este informe se tocaran 4 puntos relevantes sobre el diseño, analisis y recomendaciones que deja el modelo realizado para la compañia

1. Para el desarrollo de este modelo estadistico se realizo una division de los datos en proporcion de 70/30 para lidiar con problemas de sobreajuste a los datos del modelo lo cual mas adelante vamos a medir que sucedio con esto. en cuanto a porque se selecciono este modelo, se realizaron pruebas de 3 modelo un con las 10 variables disponibles, el segundo con 5 y el ultimo seleccionado con cuatro, los 2 primeros modelos se descartaron el primero por problemas con las variables predictoras ya que luego de hacer la prueba del P>|t| que nos indica que tan significantes o relevantes son las variables dentro del modelo y se descartaron la mitad de la variables, el segundo modelo se descarto ya que la variable ingresos mensuales hacia que el modelo tuviera problemas de multicolinealidad, y se opto por este modelo que paso las preubas de los diferentes estadisticos como R2 ajustado que indica que en un 36% de la variabilidad de las ventas pueden ser explicadas por las variables seleccionadas, una prob f statistic menor al 0.001 que demuestra una buena significancia estadistica del modelo y valor p menores a 0.05 que indican la relevancia del coeficiente y cada variable dentro del modelo

2.La division del 70/30 en los datos se hizo para utilizar el 70 para entrenar el modelo y luego testear con el 30 restante que no tuviera problemas de sobreajuste a los datos, esto se valido con el indicador del r2 el cual es muy similar en ambos demostrando que no hay un sobreajuste del modelov a los datos

3.La ecuacion del modelo quedo de la siguiente manera: ventas del cliente(y)= 159.28(a)+2.36 Edad(x1)+1.83tiempo(x2)+6.48visitas(x3)+298.87(x4)
todas las variables presentan coeficientes positivos que indican una relacion directa con las ventas o en  otras palabras el valor de las ventas aumenta en la proporcion que aumente cada una de las variables, donde el intercepto es el valor de referencia con valor de 159.28 dolares cuando las variables sean 0, para x1 por cada año adicional de edad del cliente incrementa en 2.60 dolares las ventas, x2 cada mes adicional de fidelidad aporta 1.67 dólares extra en ventas mensuales, x3 cada visita mensual adicional al sitio web genera cerca de 7.07 dólares más en ventas, x4 con el mayor coeficiente que indica que por cada vez que le aparece al cliente el descuento promedio eleva las ventas en 298.87 dólares


4.Se recomienda con esta informacion se concluye que se deberia diseñar promociones segmentadas y comunicar descuentos en canales de mayor conversión potenciará las ventas de forma inmediata, ademas de fomentar programas de retencion de clientes que aporten valor sostenido a la empresa y el uso del sitio web añadiendo contenido personalizado o notificaciones que favorezcan al crecimiento de las ventas de la compañia